client/fingerprint_manager.go

```diff
@@ -0,0 +1,122 @@
+package client
+
+import (
+	"log"
+	"time"
+
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/driver"
+	"github.com/hashicorp/nomad/client/fingerprint"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// FingerprintManager runs a client fingerprinters on a continuous basis, and
+// updates the client when the node has changed
+type FingerprintManager struct {
+	getConfig func() *config.Config
+	node      *structs.Node
+	client    *Client
```

Comment:
Instead of giving this access to the whole client struct, consider using a func type to encapsulate what this does with the client. Something like 
```
type FingerPrintResponseUpdater func(resp *structs.FingerPrintResponse)

// in client.go when you initialize the FingerPrintManager.. 
responseUpdaterFunc := func()  {
c.updateNodeFromFingerprint(&response)
}
fingerprintManager := &FingerprintManager{
 	getConfig: func() *config.Config {
 		c.configLock.Lock()
 		defer c.configLock.Unlock()
 		return c.config
 	},
 	node:   c.config.Node,
 	responseUpdater: responseUpdaterFunc,
        shutDownCh:c.shutDownch,
  	logger: c.logger,
 	}
 
```
Doing it this way lets you tighten down exactly what FingerprintManager has access to - it shouldn't need to know about the client or need the whole client struct, instead, it calls a callback function that's passed into it when fingerprint responses are available.

Also see how [AllocStateUpdater](https://github.com/hashicorp/nomad/blob/a05862dbdf154c7d046d4462237ce647b2a66d26/client/alloc_runner.go#L33) works for an example of this pattern.

Reply:
Sure, this was something I was considering, but at the time it wasn't clear how much of the client the fingerprint manager needed access to. Adding the response updater and a shutdown channel seem still low overhead, so will change that. Thanks!